src/utils/execve.ts

```diff
@@ -0,0 +1,83 @@
+import { utils, Path } from "tea"
+const { host } = utils
+
+export default function({cmd: argv, env}: {cmd: string[], env: Record<string, string>}) {
+  const filename = host().platform == 'darwin' ? '/usr/lib/libSystem.dylib' : 'libc.so.6'
+
+  const libc = Deno.dlopen(
+    filename, {
+      execve: {
+        parameters: ["pointer", "pointer", "pointer"],
+        result: "i32"
+      },
+      errno: {
+        type: "i32"
+      }
+    }
+  )
+
+  find_in_PATH(argv, env.PATH)
+
+  const command = Deno.UnsafePointer.of(new TextEncoder().encode(`${argv[0]}\0`))
+  const env_ = Object.entries(env).map(([key, value]) => `${key}=${value}\0`)
+
+  libc.symbols.execve(command, arr_to_c(argv), arr_to_c(env_))
+
+  switch (libc.symbols.errno) {
+    case 2:  //ENOENT:
+      // yes: strange behavior from execve here indeed
+      if (new Path(argv[0]).exists()) {
+        throw new Deno.errors.PermissionDenied()
+      } else {
+        throw new Deno.errors.NotFound()
+      }
+    case 13:
+      throw new Deno.errors.PermissionDenied()
+    case 7:  //E2BIG:
+    case 14: //EFAULT:
+    case 5:  //EIO:
+    case 62: //ELOOP:
+    case 63: //ENAMETOOLONG:
+    case 8:  //ENOEXEC:
+    case 12: //ENOMEM:
+    case 20: //ENOTDIR:
+    case 26: //ETXTBSY:
+      throw new Error(`execve failed (${libc.symbols.errno})`)
+  }
+}
+
+function arr_to_c(arr: string[]): Deno.PointerValue {
```

Comment:
Just wanted to reference a clean way of doing this

https://github.com/aapoalas/libclang_deno/blob/main/lib/utils.ts#L9-L46

Reply:
Linked implementation doesn't null terminate the array, but it's a pretty trivial change to do so.

Reply:
Unfortunately cannot import directly without adding a bunch of junk:

```
error: Error: Cannot load libclang without LIBCLANG_PATH environment variable
```

So I will copy pasta it ☹️

Reply:
Getting weird issues with this code:

https://github.com/teaxyz/cli/actions/runs/5436798144/jobs/9886901529?pr=621#step:7:17

I'm not really sure what to do about it, the docs for this function are absent.

https://deno.land/api@v1.34.3?s=TextEncoderEncodeIntoResult

Usually it would be a “do another write” type thing but I don’t get the issue so will just revert to the code I had before which—while being less efficient—was good enough and didn't have this issue.

Reply:
Could always be improved in a follow-up PR.